include/up-cpp/utils/Expected.h

```diff
@@ -22,77 +22,114 @@ namespace uprotocol::utils {
-
-	constexpr Expected();
-	constexpr Expected(const Expected& other);
-	constexpr Expected(Expected&& other) noexcept;
-
-	constexpr explicit Expected(T&& v);
-	constexpr explicit Expected(E&& e);
-
-	Expected& operator=(const Expected& other);
-	Expected& operator=(Expected&& other);
-
-	Expected& operator=(const T& other);
-	Expected& operator=(T&& other);
-
-	Expected& operator=(const E& other);
-	Expected& operator=(E&& other);
-
-	constexpr explicit operator bool() const noexcept;
-	constexpr bool has_value() const noexcept;
-
-	constexpr T& value() &;
-	constexpr const T& value() const&;
-	constexpr T&& value() &&;
-	constexpr const T&& value() const&&;
-
-	constexpr const E& error() const& noexcept;
-	constexpr E& error() & noexcept;
-	constexpr const E&& error() const&& noexcept;
-	constexpr E&& error() && noexcept;
-
-	constexpr T value_or(T&& default_value) const&;
-	constexpr T value_or(T&& default_value) &&;
-
-	template <typename F>
-	constexpr auto and_then(F&& f) &;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&;
-	template <typename F>
-	constexpr auto and_then(F&& f) &&;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&&;
-
-	template <typename F>
-	constexpr auto or_else(F&& f) &;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&;
-	template <typename F>
-	constexpr auto or_else(F&& f) &&;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&&;
+public:
+	constexpr Expected() { new (&yay) T(); }
+	constexpr Expected(const T& rhs) { new (&yay) T(rhs); }
+	constexpr Expected(const Unexpected<E>& rhs) : ok(false) {
+		new (&nay) E(rhs.value());
+	}
+	template <class U = T>
+	constexpr explicit Expected(U&& rhs) {
+		new (&yay) T(std::forward<U>(rhs));
+	}
+
+	constexpr Expected(const Expected& rhs) : ok(rhs.ok) {
+		if (ok)
+			new (&yay) T(rhs.yay);
+		else
+			new (&nay) E(rhs.nay);
+	}
+
+	constexpr bool has_value() const noexcept { return ok; }
+
+	constexpr explicit operator bool() const noexcept { return ok; }
+
+	template <class U>
+	constexpr T value_or(U&& v) const& noexcept {
+		return ok ? yay : v;
+	}
+
+	constexpr T value() const {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to access value() when unexpected.");
+		return yay;
+	}
+
+	constexpr const E& error() const {
+		if (ok)
+			throw BadExpectedAccess(
+			    "Attempt to access error() when not unexpected.");
+		return nay;
+	}
+
+	constexpr T& operator*() {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to non-const dereference expected value when "
+			    "unexpected.");
+		return yay;
+	}
+
+	constexpr const T& operator*() const {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to const dereference expected value when unexpected.");
+		return yay;
+	}
+
+	constexpr T* operator->() {
+		if (!ok)
+			throw BadExpectedAccess(
+			    "Attempt to dereference expected pointer when unexpected.");
+		return &**this;
+	}
 
 private:
+	union {
+		T yay;
+		E nay;
+	};
```

Comment:
Ah... I see why this is not likely to leak memory. A `union` doesn't allow non-trivial destructors, so it (probably) doesn't matter if `T`'s destructor runs. However, that does not meet our needs for `Expected`. We need to be able to return `T`'s with non-trivial destructors, like `std::expected` does.

You can see an example of this in `UTransport::registerListener()` - it is returning `Expected<ListenHandle, v1::UStatus>`.

As soon as the union is removed to allow types with non-trivial destructors, the memory leaks will appear:
![image](https://github.com/eclipse-uprotocol/up-cpp/assets/10570177/c810b315-30c5-4dfa-b38a-c864b57ec4f9)
(note: I removed the union and added a test case with a struct that contains a `unique_ptr` that is allocated in the struct's constructor so that the leak could be seen)

Reply:
Maybe a `std::variant<T, E>` would help. It already has most of the necessary logic and implementations around moving / constructing / copying values, and checking the type contained could take the place of `bool ok`.